Add non-resource-request support to ABAC, fix kubectl for namespaced users

liggitt opened the pull request:
Rebased and built on https://github.com/kubernetes/kubernetes/issues/13667
- authorizer.Attributes
  - Added attribute indicating whether the request was for a resource (`IsResourceRequest`)
  - Added attribute containing the request path (can be used in combination with `IsResourceRequest` to authorize non-resource requests to paths like `/swaggerapi`, or `/apis` or `/healthz`, etc)
- ABAC policy
  - Added support for matching on API groups attribute
  - Added support for matching non-resource requests (fixes #13097)
  - Added line number and context to load errors
  - Ignore empty and comment lines
  - Converted to versioned resources (similar to scheduler policy objects) (fixes #15117)
  - In "abac.authorization.kubernetes.io/v1beta1", explicitly require `"user":"*","namespace":"*","resource":"*","apiGroup":"*","nonResourcePath":"*"` to match all (deny by default, rather than grant by default)
  - Defined a "abac.authorization.kubernetes.io/v0" version for ABAC policy rule objects with conversions to "abac.authorization.kubernetes.io/v1beta1" preserving current behavior (preserved all ABAC tests as well)

liggitt commented:
@deads2k @mkulke PTAL
@mkulke can you comment to appease the cla bot?

liggitt commented:
@kubernetes/kube-iam this should give us sufficient authz power to allow API discover and version negotiation while still constraining users to namespaces. I'd like to update the ABAC file format to a versioned format (falling back to the current JSONL format on deserialize errors for backwards compatibility), but will do that in another PR

mkulke commented:
I signed it!

deads2k commented:
@liggitt the sticking point last time was whether non-resource access was default allow or default deny.  Which did you choose?

beni05 commented:
nem-erőforrás ,

mkulke commented:
@liggitt LGTM

i built and tested the branch, seems to work fine:

    $ ../../tmp/kubernetes/_output/dockerized/bin/linux/amd64/kubectl version
    Client Version: version.Info{Major:"1", Minor:"2+", GitVersion:"v1.2.0-alpha.1.1513+e162b71ae80901", GitCommit:"e162b71ae809019a631c0704f18c3a1de631fb86", GitTreeState:"clean"}
    Server Version: version.Info{Major:"1", Minor:"2+", GitVersion:"v1.2.0-alpha.1.1513+e162b71ae80901", GitCommit:"e162b71ae809019a631c0704f18c3a1de631fb86", GitTreeState:"clean"}

    $ ../../tmp/kubernetes/_output/dockerized/bin/linux/amd64/kubectl --context=orange run nginx --image=nginx
    replicationcontroller "nginx" created

    $ ../../tmp/kubernetes/_output/dockerized/bin/linux/amd64/kubectl --context=black run nginx --image=nginx
    replicationcontroller "nginx" created

    $ ../../tmp/kubernetes/_output/dockerized/bin/linux/amd64/kubectl --context=orange run nginx --image=nginx
    replicationcontroller "nginx" created

    $ ../../tmp/kubernetes/_output/dockerized/bin/linux/amd64/kubectl --context=black get po
    NAME          READY     STATUS    RESTARTS   AGE
    nginx-b80ks   1/1       Running   0          5m

    $ ../../tmp/kubernetes/_output/dockerized/bin/linux/amd64/kubectl --context=liggitt --all-namespaces=true get po
    NAMESPACE   NAME          READY     STATUS    RESTARTS   AGE
    black       nginx-b80ks   1/1       Running   0          6m
    orange      nginx-ayprx   1/1       Running   0          6m

mkulke commented:
@liggitt : about the bot: anything specific it needs to read here?

liggitt commented:
not sure, try @ mentioning the bot and saying you're okay with these being contributed to Google... no idea what magic incantation it expects

mkulke commented:
@googlebot i signed it!

liggitt commented:
flakes, only one that looked like it could be relevant (kubectl test) passed locally:
```
[Fail] Kubectl client Kubectl describe [It] should check if kubectl describe prints relevant information for rc and pods [Conformance] 
/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/test/e2e/kubectl.go:936

[Fail] Probing container [It] with readiness probe should not be ready before initial delay and never restart [Conformance] 
/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/test/e2e/container_probe.go:59

[Fail] Pods [It] should have their container restart back-off timer increase exponentially 
/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/test/e2e/pods.go:859
```

@k8s-bot test this

beni05 commented:
ellenőriznie kell,

liggitt commented:
@mkulke try the same thing you put in the other PR: "Signed the CLA, the company name is "thinkstep AG"."

mkulke commented:
Signed the CLA, the company name is "thinkstep AG".

mkulke commented:
@liggitt: it could be that the bot is confused about the shuffling of PRs, someone in slack noticed the same behaviour.

liggitt commented:
maybe... do you care if I squash to a single commit?

beni05 commented:
minden rendben ,

liggitt commented:
> @deads2k the sticking point last time was whether non-resource access was default allow or default deny. Which did you choose?

I chose default deny, because default allow seems crazy when a field name typo in the policy file means the rule would be empty, allowing all. Granted, that's how namespace and resource behave now... but I was hoping to avoid propagating that.

Thinking further, though, `{"user":"admin"}` has to grant access to everything (resource and non-resource) to preserve compatibility. I'll add tests to make sure that doesn't break.

To get this looking the way I want internally, I may bite the bullet and add a versioned schema for ABACPolicy, and start requiring "*" for namespace and resource to match. Something like this:
```
{
  "apiVersion":"v1",
  "kind":"ABACPolicy",
  "rules":[
    {"user":"admin","namespace":"*",resource:"*"},
    {"user":"admin","nonResourcePath":"*"},
  ]
}
```

Deserialize failures on old ABAC files could attempt to read the JSONL format and convert to the versioned form like this:
```
{"user":"u"}                  -> {"user":"u","nonResourcePath":"*"}, {"user":"u","namespace":"*","resource":"*"}

{"user":"u","namespace":"ns"} -> {"user":"u","namespace":"ns","resource":"*"}

{"user":"u","resource":"r"}   -> {"user":"u","namespace":"*","resource":"r"}
```

Before I start on that, any objections from @kubernetes/kube-iam on the approach?

beni05 commented:
a bot zavaros a csoszogós és Merge,?

beni05 commented:
Gondolkodás további,("admin", "névtér": "*", az erőforrás-: "*"},

liggitt commented:
or, while bullets are being bitten, this:
```
{
  "apiVersion":"v1",
  "kind":"ABACPolicyList",
  "items":[
    {
      "user":"u",
      "groups":["g1","g2"],
      "verbs":["get","list","watch"], // or ["*"]
      "namespaces":["ns1","ns2"], // or ["*"]
      "resources":["pods","replicationcontrollers"], // or ["*"]
      "nonResourcePaths":["/api","/api/*","/apis","/apis/*"] // or ["*"]
    },
    ...
  ]
}
```
Being able to grant a user or group permission across namespaces, verbs, and resources in a single rule improves readability quite a bit. Depends on how real we envision this being, and whether we see this as a progression toward dynamic authz.

deads2k commented:
> or, while bullets are being bitten, this:

No.  That makes `verbs` for `nonResourcePaths` weird.

beni05 commented:
Thenthme,

mkulke commented:
@liggitt: sure, go ahead.

beni05 commented:
"apiVersion": "V1",
  "fajta": "ABACPolicyList",
  "példány": [
    {
      "felhasználó": "u",

beni05 commented:
GitHub offers free accounts for users and organizations working on open ..!

deads2k commented:
I like default denies and other than allowing both non-resource and resource rules in a single statement, I like the format of https://github.com/kubernetes/kubernetes/pull/16148#issuecomment-150631961.

If you want a new format, would it be easier to simply have a second authorizer to implement it?

yuvipanda commented:
/cc @erictune (This is the kubectl namespacing patch I was talking about)

liggitt committed:
Add non-resource and API group support to ABAC authorizer, version ABAC policy rules

liggitt commented:
@erictune updated, PTAL

deads2k commented on pkg/auth/authorizer/abac/api/register.go:
```diff
@@ -0,0 +1,32 @@
+var Scheme = runtime.NewScheme()
+
+func init() {
+	Scheme.AddKnownTypes("",
```
If you're going to do this, I'd like them in a different group.

liggitt replied:
added a second commit with an API group... not sure what to call the group

deads2k commented:
We allow dots.  Maybe `abac.kubernetes.io` ?

liggitt commented:
This PR doesn't really set up the objects to be consumable by the API yet (no ObjectMeta, etc). Also, I wasn't sure how real we wanted the group to be... e.g. "authz.kubernetes.io", kind of "ABACPolicy"?

erictune commented on docs/admin/authorization.md:
```diff
@@ -99,21 +111,35 @@ A request has attributes which correspond to the properties of a policy object.
-To permit any user to do something, write a policy with the user property unset.
-To permit an action Policy with an unset namespace applies regardless of namespace.
+To permit any user to do something, write a policy with the user property set to "*".
+To permit a user to do anything, write a policy with the namespace, resource, and nonResourcePath properties set to "*".
```
This is a non-backwards compatible change?

liggitt replied:
see behavior in the load from file... resource is now tagged with kind/version. "v1" has this behavior. old rules are converted to preserve behavior

liggitt replied:
all existing files will continue to work exactly as they do today. tests ensure "v0" rules are converted and behave identically

erictune commented on docs/admin/authorization.md:
```diff
@@ -99,21 +111,35 @@ A request has attributes which correspond to the properties of a policy object.
- 2. Kubelet can read any pods: `{"user":"kubelet", "resource": "pods", "readonly": true}`
- 3. Kubelet can read and write events: `{"user":"kubelet", "resource": "events"}`
- 4. Bob can just read pods in namespace "projectCaribou": `{"user":"bob", "resource": "pods", "readonly": true, "namespace": "projectCaribou"}`
+ 1. Alice can do anything to all resources:                  `{"apiVersion": "abac/v1", "kind": "Policy", "user": "alice", "namespace": "*", "resource": "*", "resourceGroup": "*"}`
```
Will the old way also work?  It needs to since people use it.

liggitt replied:
all existing files will continue to work as they do today. Once you add `"kind":"Policy","apiVersion":"abac.authz.kubernetes.io/v1",...` to a rule, it has to match v1 semantics, which is deny by default

erictune commented on docs/admin/authorization.md:
```diff
@@ -79,18 +81,28 @@ The file format is [one JSON object per line](http://jsonlines.org/).  There sho
+    - `namespace`, type string; a namespace string. `*` matches all resource requests.
+
+  - Non-resource-matching properties:
+  - `nonResourcePath`, type string; matches the non-resource request paths (like `/version` and `/apis`). `*` matches all non-resource requests. `/foo/*` matches `/foo/` and all of its subpaths.
```
For API resources, can we call this `apiGroup`, and imply the /api or /apis prefix?  For miscellaneous resources, keep `nonResourcePath`, and forbid specifying both `apiGroup` and `nonResourcePath`?

liggitt replied:
Not sure what you mean by this... how would you express that user "X" could request "/apis" (to do discovery)?

erictune replied:
In particular, how about the Attributes does not include path for an API resource.  This allows us to change the path format in the future without invalidating ABAC objects.

liggitt replied:
ABAC isn't doing anything with the path attribute for resource requests, only for non-resource ones...

deads2k commented on pkg/auth/authorizer/abac/abac.go:
```diff
@@ -79,62 +64,150 @@ func NewFromFile(path string) (policyList, error) {
+	if !a.IsResourceRequest() {
+		return false
+	}
+	if p.Namespace != "*" && p.Namespace != a.GetNamespace() {
```
rewrite to have positive logic and return true?

deads2k replied:
> rewrite to have positive logic and return true?

Oh, I see, structured backward of downstream.  Ok.

liggitt replied:
returning true requires nesting... false lets us exit early... I can swap if you want

deads2k replied:
> returning true requires nesting... false lets us exit early... I can swap if you want

I think of authorization evaluation in terms of successively tighter checks, so nesting makes more sense to me.  "Match this and this and this and this".  Having seen both, you prefer this?

liggitt replied:
I could see either way, updated

deads2k commented on pkg/auth/authorizer/abac/abac_test.go:
```diff
@@ -320,27 +620,282 @@ func TestPolicy(t *testing.T) {
+				ResourceRequest: false,
+			},
+			matches: true,
+			name:    "v1 non-resource wildcard subpath match",
```
I'm sure there's a test missing here somewhere. :)

deads2k commented on pkg/auth/authorizer/abac/api/register.go:
```diff
@@ -0,0 +1,32 @@
+var Scheme = runtime.NewScheme()
+
+func init() {
+	Scheme.AddKnownTypes("abac.authz.kubernetes.io/",
```
why `authz` instead of `authorization`?

liggitt replied:
you don't think it's long enough already?

liggitt replied:
lengthened

deads2k commented:
I'd prefer to spend the letters and avoid `authz` in the new group.

Other than that, lgtm.

erictune commented:
@yuvipanda this PR is redesigning the ABAC file format.  Since you are a known user of that format, your feedback is desired.

erictune commented:
Can you explain the thinking behind:

> In "abac.authz.kubernetes.io/v1", explicitly require "user":"*" or "namespace":"*" or "resource":"*" or "resourceGroup":"*" to match all (deny by default, rather than grant by default)

liggitt commented:
right now, the most dangerous policy you can write is `{}`, which allows anyone do to anything. A typo in an attribute name is similarly dangerous: `{"username":"bob","ns":"bobs-ns"}` actually allows anyone to do anything. Taking the absence of an attribute to mean "allow all" makes it very hard to look at a policy and know what it is granting.

liggitt commented:
@yuvipanda the `pkg/auth/authorizer/abac/example_policy_file.jsonl` in this PR is a good example of what it would look like. The proposed changes are:
1. require kind/apiVersion be specified
2. require attributes intended to allow all specify `...="*"`. The policy to allow all users to do everything would look like this:
```
{"apiVersion": "abac.authorization.kubernetes.io/v1", "kind": "Policy", "user":"*","namespace": "*","resource": "*","resourceGroup": "*"}
```

erictune commented:
okay, my questions are answered.   thanks for doing this!

liggitt commented:
squashed. @mkulke, your feedback is also appreciated

erictune commented on pkg/auth/authorizer/abac/api/v1/types.go:
```diff
@@ -0,0 +1,54 @@
+
+	// ResourceGroup is the name of an API group. ResourceGroup, Resource, and Namespace are required to match resource requests.
+	// "*" matches all API groups
+	ResourceGroup string `json:"resourceGroup,omitempty"`
```
@lavalamp @bgrant0607 any objection to use of the term `ResourceGroup` for API group.

liggitt replied:
Honestly, I'm not a fan either, but `apiGroup` seemed too easily confused with the `apiVersion` field of the policy object itself (which contains a group). Suggestions welcome.

lavalamp replied:
IMO, APIGroup would be *much* clearer.

It's extremely unfortunate that our field is named APIVersion when in fact it's actually APIGroupVersion, but I think avoiding the term APIGroup and making up something new makes that problem worse, not better.

deads2k replied:
> IMO, APIGroup would be much clearer.
> 
> It's extremely unfortunate that our field is named APIVersion when in fact it's actually APIGroupVersion, but I think avoiding the term APIGroup and making up something new makes that problem worse, not better.

The choice of `ResourceGroup` makes it clear that the field is about the resource you're writing the rule for, not a field related the `Policy` object itself.

lavalamp replied:
Oh, I see there's an inline TypeMeta. I agree, that is confusing. To follow our other API objects, I think most of these fields should be put in a "PolicySpec" type, which would avoid the confusion.

yuvipanda commented:
I like it! If you consider this stable enough I'm happy to cherry-pick this patch and put it on my cluster to see what happens :)

I'm curious why JSONL is being used here instead of just JSON.

deads2k commented:
This rest of this week is a US holiday, so I'll leave this until next week for comments.

liggitt commented:
> I'm curious why JSONL is being used here instead of just JSON. 

mostly for the ability to go rule by rule and decide whether it needs converting for backwards-compatibility. I could make a normal PolicyList object with an `items` field, and if that decode fails, fall back to line-by-line conversion. I don't feel strongly either way. Before this file could actually be submitted to the API, there would be other work required as well (each rule would need a unique name, which is sort of annoying)

mkulke commented:
I like JSONL at this particular place, it allows you to easily concatenate groups of policies w/o having to resort jq or something else. We have a bunch of policies sitting in a folder (some are immutable, because they've put there by cloud-init) and our apiserver systemd unit cat's them into a single file before starting the apiserver.

liggitt commented:
Unrelated failure:
```
Downward API volume [It] should update labels on modification [Conformance] 
/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/test/e2e/downwardapi_volume.go:106
```
@k8s-bot test this

mkulke commented:
@liggitt: i spinned up a cluster with bins from this PR the apiserver seems to swallow the v1.0 policy file, but I noticed the following behaviour:

On apiserver:

```
k8s-mgns-master ~ # systemctl cat kube-apiserver
# /etc/systemd/system/kube-apiserver.service
[Unit]
Description=Kubernetes API Server
Documentation=https://github.com/GoogleCloudPlatform/kubernetes
Requires=k8s-etcd.service
After=k8s-etcd.service

[Service]
ExecStartPre=/usr/bin/wget -N -P /opt/bin https://<removed>
ExecStartPre=/usr/bin/chmod +x /opt/bin/kube-apiserver
ExecStartPre=-/usr/bin/mkdir -p /var/lib/kubernetes
ExecStartPre=/bin/sh -c 'cat /opt/k8s/authn/* > /var/lib/kubernetes/authn.csv'
ExecStartPre=/bin/sh -c 'cat /opt/k8s/authz/* > /var/lib/kubernetes/authz.jsonl'
ExecStart=/opt/bin/kube-apiserver \
--cluster-name=k8s-mgns \
--secure-port=443 \
--service-cluster-ip-range=10.100.0.0/16 \
--etcd_servers=http://127.0.0.1:3379 \
--admission_control=NamespaceLifecycle,NamespaceExists,LimitRanger,SecurityContextDeny,ResourceQuota \
--token-auth-file=/var/lib/kubernetes/authn.csv \
--tls-cert-file=/opt/k8s/server.crt \
--tls-private-key-file=/opt/k8s/server.key \
--authorization_mode=ABAC \
--authorization_policy_file=/var/lib/kubernetes/authz.jsonl \
--logtostderr=true
Restart=always
RestartSec=10
```

```
k8s-mgns-master ~ # cat /var/lib/kubernetes/authn.csv | grep orange
59a71785418649309fdd27f55570dc95,orange,orange
```

```
k8s-mgns-master ~ # cat /var/lib/kubernetes/authz.jsonl | grep orange
{"user": "orange", "namespace": "orange"}
```

```
$ http https://k8s-mgns-api.thinkstep.io/api/v1/namespaces/default/pods "Authorization: Bearer 59a71785418649309fdd27f55570dc95"
HTTP/1.1 200 OK
Content-Length: 130
Content-Type: application/json
Date: Thu, 26 Nov 2015 15:48:10 GMT

{
    "apiVersion": "v1", 
    "items": [], 
    "kind": "PodList", 
    "metadata": {
        "resourceVersion": "715", 
        "selfLink": "/api/v1/namespaces/default/pods"
    }
}
```

The request bears the user `orange`, which is restricted to the the namespace `orange`, but I can read objects from the `default` namespace.

expected  behaviour? this might be surprising for users.

liggitt commented:
Can you include the entire authz file? Policy rules are additive… I would assume another rule is allowing that access.

mkulke commented:
@liggitt 

sure:

```
k8s-mgns-master ~ # cat /var/lib/kubernetes/authz.jsonl              
{"user": "admin"}
{"user": "scheduler", "readonly": true, "resource": "pods"}
{"user": "scheduler", "resource": "bindings"}
{"user": "kubelet", "resource": "pods"}
{"user": "kubelet", "readonly": true, "resource": "services"}
{"user": "kubelet", "readonly": true, "resource": "endpoints"}
{"user": "kubelet", "resource": "events"}
{"user": "kubelet", "resource": "nodes"}
{"user": "kube-proxy", "readonly": true, "resource": "services"}
{"user": "kube-proxy", "readonly": true, "resource": "endpoints"}
{"user": "kube-proxy","resource": "events"}
{"readonly": true, "nonResourcePath": "/api"}
{"readonly": true, "nonResourcePath": "/version"}
{"readonly": true, "nonResourcePath": "/swaggerapi/api/v1"}
{"readonly": true, "nonResourcePath": "/swaggerapi/experimental/v1"}
{"user": "orange", "namespace": "orange"}
```

liggitt commented:
Also, the example rule you showed is still the unversioned policy format, which is backward compatible, meaning omitted attributes allow all access. Did you mean to test with the v1 format?

liggitt commented:
`nonResourcePath` only exists in v1. Your nonResourcePath rules are read as `{readonly:true}`, which allows read access to everything to everyone. I could add validation for unversioned legacy policy rules and reject ones with unrecognized attributes to prevent that from happening...

liggitt commented:
That validation would probably be good anyway to prevent attribute typos from accidentally allowing too much (that problem is limited to the legacy rules)

mkulke commented:
@liggitt you're right, did not consider this. removing those `nonResourcePath` policies does indeed make  namespace isolation work:

```
$ http https://k8s-mgns-api.thinkstep.io/api/v1/namespaces/default/pods "Authorization: Bearer 59a71785418649309fdd27f55570dc95"
HTTP/1.1 403 Forbidden
Content-Length: 44
Content-Type: text/plain; charset=utf-8
Date: Thu, 26 Nov 2015 16:01:16 GMT

Forbidden: "/api/v1/namespaces/default/pods"
```

and as @yuvipanda and me are probably the only ones with `nonResourcePath` policies, it's negligable I guess.

mkulke commented on docs/admin/authorization.md:
```diff
@@ -99,21 +111,35 @@ A request has attributes which correspond to the properties of a policy object.
+ 2. Kubelet can read any pods:                               `{"apiVersion": "abac.authorization.kubernetes.io/v1", "kind": "Policy", "user": "kubelet", "namespace": "*", "resource": "pods", "readonly": true}`
+ 3. Kubelet can read and write events:                       `{"apiVersion": "abac.authorization.kubernetes.io/v1", "kind": "Policy", "user": "kubelet", "namespace": "*", "resource": "events"}`
+ 4. Bob can just read pods in namespace "projectCaribou":    `{"apiVersion": "abac.authorization.kubernetes.io/v1", "kind": "Policy", "user": "bob", "namespace": "projectCaribou", "resource": "pods", "readonly": true}`
+ 5. Anyone can make read-only requests to all non-API paths: `{"apiVersion": "abac.authorization.kubernetes.io/v1", "kind": "Policy", "user": "*", "readonly": "true", "nonResourcePath": "*"}`
```
quotes, has to be: `"readonly": true`

mkulke commented:
@liggitt: thanks, LGTM:

```
k8s-mgns-master ~ # cat /var/lib/kubernetes/authz.jsonl | grep -e nonResourcePath -e orange
{"apiVersion": "abac.authorization.kubernetes.io/v1", "kind": "Policy", "user": "*", "readonly": true, "nonResourcePath": "*"}
{"user": "orange", "namespace": "orange"}
```

```
kubectl --context=mgns-orange run nginx --image=nginx # using user orange
CONTROLLER   CONTAINER(S)   IMAGE(S)   SELECTOR    REPLICAS
nginx        nginx          nginx      run=nginx   1
```

erictune commented:
Fixes #13097

liggitt commented:
I think @deads2k was waiting until tomorrow for feedback before tagging. I plan to add validation around legacy rules as a follow up (to catch attribute typos or unsupported attributes accidentally granting broader permissions than intended)

yuvipanda commented:
/me agrees with @mkulke that we two are probably the only ones carrying the older version of this patch, so validation for that particular version of nonResourcePath is probably unnecessary.

I'll wait for this patch to land and then roll it unto our deployment setup!

liggitt commented:
> /me agrees with @mkulke that we two are probably the only ones carrying the older version of this patch, so validation for that particular version of nonResourcePath is probably unnecessary.

Sure, but I've seen other cases where someone was trying to limit a user to a namespace and set `"ns":"somenamespace"`, which silently got dropped and actually granted the user full access to everything. I'll do the validation as a follow up so as not to block this PR

deads2k commented:
lgtm and feedback looks positive.

liggitt commented:
```
• Failure in Spec Setup (BeforeEach) [48.410 seconds]
Networking
/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/test/e2e/networking.go:249
  should provide Internet connection for containers [Conformance] [BeforeEach]
  /go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/test/e2e/networking.go:82

  Dec  1 05:56:14.902: Unable to connect/talk to the internet: Get http://google.com: dial tcp 74.125.202.139:80: i/o timeout
  /go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/test/e2e/networking.go:47
```
@k8s-bot test this

lavalamp commented:
Does this include API types? It appears to. The directory structure is not right.

Please put the api types in `pkg/apis/<yourgroup>/...`. If you want to keep the types in with your own code, that's inconsistent with the rest of our codebase. I can live with it, but you still need to keep the api directory structure consistent: `path/to/your/base/apis/<yourgroup>/...`. Note the `s` in apis. If you do the latter, you'll probably have to fix a bunch of tools.

...I need to update my proposal doc so I have something official to point people at, sigh.

deads2k commented:
@liggitt untagged so you can answer comments

deads2k commented:
> Does this include API types? It appears to. The directory structure is not right.

@lavalamp These types are sort of "side-loaded" from file.  Their location is historical and I don't think we'll want to expose them as resources.

liggitt commented:
> Does this include API types? It appears to. The directory structure is not right.
>
> Please put the api types in pkg/apis/<yourgroup>/.... If you want to keep the types in with your own code, that's inconsistent with the rest of our codebase. I can live with it, but you still need to keep the api directory structure consistent: path/to/your/base/apis/<yourgroup>/.... Note the s in apis. If you do the latter, you'll probably have to fix a bunch of tools.

This is an incremental change to make the ABAC policy types versioned and typed (just like `scheduler/api/types.go`), and handle migrating from the old untyped unversioned files.

More work would be required to make them ready to expose as resources (like adding ObjectMeta and a Name field, etc). That might be valuable at some point, but that's beyond the scope of this PR

lavalamp commented:
> This is an incremental change to make the ABAC policy types versioned and typed (just like scheduler/api/types.go), and handle migrating from the old untyped unversioned files.

OK, I understand. Sorry, but I still would like the directory structure changed, since it is really new.

You might consider calling v1 v1beta1 instead.

liggitt commented:
> Please put the api types in `pkg/apis/<yourgroup>/...`

Go packages can't have dots in them...

liggitt commented:
Moved under `/apis`, named subfolder `abac` for lack of a closer name, changed `v1` to `v1beta1`.

I also added `spec` and renamed `resourceGroup` back to `apiGroup`. Not sure how I feel about spec nesting, but I guess it works.

liggitt commented:
will wait to squash until we decide we like this direction

lavalamp commented:
Thank you for humoring me, no more objections here!

deads2k commented:
lgtm.

Squash and tag at will.

liggitt commented:
@lavalamp thinking further, was there a specific reason for making the version `v1beta1` rather than `v1`? It's a very granular api group, so bumping versions wouldn't be hard, I just wondered

lavalamp commented:
Yeah-- if we call it v1beta1 then I don't have to be picky about the
contents of the API types :) :)

On Wed, Dec 2, 2015 at 10:37 AM, Jordan Liggitt <notifications@github.com>
wrote:

> @lavalamp <https://github.com/lavalamp> thinking further, was there a
> specific reason for making the version v1beta1 rather than v1? It's a
> very granular api group, so bumping versions wouldn't be hard, I just
> wondered
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/16148#issuecomment-161392546>
> .
>

liggitt commented:
rebased
